src/System.Security.Principal.Windows/src/System/Security/Principal/WindowsImpersonationContext.cs

```diff
@@ -0,0 +1,70 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.ComponentModel;
+using System.Threading;
+using Microsoft.Win32.SafeHandles;
+
+namespace System.Security.Principal
+{
+    public class WindowsImpersonationContext : IDisposable
+    {
+        private readonly SafeAccessTokenHandle _previousUserToken = SafeAccessTokenHandle.InvalidHandle;
+
+        private WindowsImpersonationContext() { }
+
+        internal WindowsImpersonationContext(SafeAccessTokenHandle currentUserToken, bool isImpersonating)
+        {
+            if (currentUserToken.IsInvalid)
+            {
+                throw new ArgumentException(SR.Argument_InvalidImpersonationToken);
+            }
+
+            if (isImpersonating)
+            {
+                // if we're in a impersonated context we need to duplicate current token to allow revert to correct identity 
```

Comment:
The _previousUserToken field is always initialized to a new object (`SafeAccessTokenHandle.InvalidHandle` returns new every time).  Either dispose it here before reassignment, or restructure the field to be set in an else here.